letsencrypt/client/dns_authenticator.py

```diff
@@ -0,0 +1,261 @@
+"""DNS Authenticator"""
+import dns.query
+import dns.tsigkeyring
+import dns.update
+import dns.resolver
+
+import zope.interface
+
+from letsencrypt.acme import challenges
+
+from letsencrypt.client import achallenges
+from letsencrypt.client import constants
+from letsencrypt.client import errors
+from letsencrypt.client import interfaces
+
+def find_valid_key(tsig_keys, domain):
+    """Search provided TSIG key pairs.
+
+    Search the keypairs in `config.dns_tsig_keys` for one valid for the
+    provided `domain`.
+
+    :param list tsig_keys: List of tuples containing key name, key secret,
+        and domains key is valid for.
+    :param str domain: Domain name to look for a key pair for.
+
+    :returns: Keyring valid for `domain` or `None`.
+    :rtype: dns.tsigkeyring
+
+    """
+    for keypair in tsig_keys:
+        if domain in keypair[2]:
+            return dns.tsigkeyring.from_text(
+                {keypair[0]: keypair[1]})
+
+def add_record(zone, token, keyring):
+    """Add record DNS request generator.
+
+    Create a `dns.message.Message` object asking the server to
+    create a TXT record for the challenge subdomain
+    (`_acme-challenge.example.com`) signed by the provided
+    TSIG keyring.
+
+    :param str zone: Zone (domain) in which the record should
+        be provisioned.
+    :param str token: Token provided by ACME server.
+    :param dns.tsigkeyring keyring: TSIG keyring object containg TSIG
+        key pair valid for the provided zone.
+
+    :returns: Add record DNS message.
+    :rtype: dns.message.Message
+
+    """
+    challenge_subdomain = "_acme-challenge.%s" % (zone)
```

Comment:
I would put the `_acme-challenge.{0}` as constant in `letsencrypt.acme.challenges.DNS.TXT_RECORD_TEMPLATE` and create

```python
@classmethod
def txt_subdomain(cls, zone):
    """Subdomain for the TXT record used for validation."""
    return cls.DNS.TXT_RECORD_TEMPLATE.format(zone)
```